experimental/oak_async/src/grpc.rs

```diff
@@ -129,3 +145,116 @@ impl<T> Drop for MultiWriter<T> {
+}
+
+// Unpin is needed to be able to peek at the first request before passing it
+// on to the handler.
+impl<T: Decodable> Unpin for GrpcRequestStream<T> {}
+
+// This should only be used by the Oak service generator. Not a stable API.
+#[doc(hidden)]
+impl GrpcRequestStream<()> {
+    pub fn new(inner: ChannelReadStream<GrpcRequest>) -> GrpcRequestStream<()> {
+        GrpcRequestStream {
+            inner: inner.peekable(),
+            is_drained: false,
+            _msg: core::marker::PhantomData,
+        }
+    }
+
+    // Casts this stream into the given message type.
+    pub fn into_type<T: Decodable>(self) -> GrpcRequestStream<T> {
+        GrpcRequestStream {
+            inner: self.inner,
+            is_drained: self.is_drained,
+            _msg: core::marker::PhantomData,
+        }
+    }
+}
+
+// This should only be used by the Oak service generator. Not a stable API.
+#[doc(hidden)]
+impl<T: Decodable> GrpcRequestStream<T> {
+    pub fn peek(self: Pin<&mut Self>) -> Peek<'_, ChannelReadStream<GrpcRequest>> {
+        Peekable::peek(Pin::new(&mut Pin::into_inner(self).inner))
+    }
+
+    pub async fn first(mut self) -> Result<T, OakError> {
+        // Unwrapping is safe here, as the gRPC stream must have at least one incoming request
+        self.next().await.unwrap()
+    }
+}
+
+impl<T: Decodable> Stream for GrpcRequestStream<T> {
+    type Item = Result<T, OakError>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if self.is_drained {
+            return Poll::Ready(None);
+        }
+        let self_ = Pin::into_inner(self);
+        let inner = Pin::new(&mut self_.inner);
+        Peekable::poll_next(inner, cx).map(|ready| {
+            ready.map(|some| {
+                some.and_then(|request| {
+                    self_.is_drained = request.last;
+                    let message = Message {
+                        bytes: request.req_msg,
+                        handles: Vec::new(),
+                    };
+                    T::decode(&message)
+                })
+            })
+        })
+    }
+}
+
+// This should only be used by the Oak service generator. Not a stable API.
+#[doc(hidden)]
+pub async fn invocation_to_requests_and_writer(
+    invocation: Invocation,
+) -> Result<(String, GrpcRequestStream<()>, ChannelResponseWriter), OakError> {
+    let receiver = invocation.receiver.ok_or_else(|| {
+        error!("Invocation did not have a receiver");
+        OakError::OakStatus(OakStatus::ErrInvalidArgs)
+    })?;
+    let mut requests = GrpcRequestStream::new(receiver.receive_stream());
+    let requests_pinned = core::pin::Pin::new(&mut requests);
+    let sender = invocation.sender.ok_or_else(|| {
+        error!("Invocation did not have a sender");
+        OakError::OakStatus(OakStatus::ErrInvalidArgs)
+    })?;
+    let response_writer = ChannelResponseWriter::new(sender);
+
+    let req1 = match requests_pinned
```

Comment:
`req1` is only used for getting the `method_name` from it. I think it is more readable to merge this and the next statement, and only declare `method_name`.